src/kde/kde.i.hh

```diff
@@ -18,48 +18,113 @@ namespace rtt_kde {
 /*!
  * \brief
- * DEFAULT reconstruction to return error if instantiation is not found
+ * epan_kernel basis function used during reconstruction
  *
- * \tparam coord enumerated value to specify KDE coordinate system
- * \tparam dim integer specifying the data dimensionality 
- * \param[in] distribution 
- * \param[in] position
- * \param[in] one_over_band_width
- * \param[in] domain_decomposed
- * \return final local function distribution
+ * Epanechnikov kenrel to be used in reconstrtuction
  *
+ * \param[in] x from kernel origin
+ * \return distribution weight based on distance from the kernel center 
+ *
+ * Test of kde.
  */
+inline double kde::epan_kernel(const double x) const {
+  const double x2 = x * x;
+  return x2 > 1.0 ? 0.0 : 0.75 * (1.0 - x2);
+}
 
-template <int coord>
-template <int dim>
-std::vector<double>
-kde<coord>::reconstruction(const std::vector<double> & /*distribution*/,
-                           const std::vector<std::array<double, 3>> & /*position*/,
-                           const std::vector<std::array<double, 3>> & /*one_over_band_width*/,
-                           const bool /*domain_decomposed*/) const {
-
-  Insist(false, "kde::reconstruction has not been implemented for this coordinate system and or "
-                "dimension combination");
-  return std::vector<double>(1, 0.0);
+//------------------------------------------------------------------------------------------------//
+/*!
+ * \brief
+ * log_transform pushes the distribution to log space which is less sensitive to sharp derivatives.
+ *
+ *
+ * \param[in] value of original distribution
+ * \param[in] bias used to ensure positivity
+ *
+ * \return the logarithmic transform of the original value
+ *
+ * Test of kde.
+ */
+inline double kde::log_transform(const double value, const double bias) const {
+  Require(value + bias > 0.0);
+  return log(value + bias);
 }
 
 //------------------------------------------------------------------------------------------------//
 /*!
  * \brief
- * epan_kernel basis function used during reconstruction
+ * log_inv_transform bring the distribution back to the original space.
  *
- * Epanechnikov kenrel to be used in reconstrtuction
  *
- * \param[in] x from kernel origin
- * \return distribution weight based on distance from the kernel center 
+ * \param[in] log_value of original distribution
+ * \param[in] bias used to ensure positivity
+ *
+ * \return the logarithmic transform of the original value
  *
  * Test of kde.
  */
-template <int coord> double kde<coord>::epan_kernel(const double x) const {
-  const double x2 = x * x;
-  return x2 > 1.0 ? 0.0 : 0.75 * (1.0 - x2);
+
+inline double kde::log_inv_transform(const double log_value, const double bias) const {
+  return exp(log_value) - bias;
 }
 
+//! Lambda to calculate a vector
+auto calc_vec = [](const auto &v1, const auto &v2) {
+  return std::array<double, 3>{v2[0] - v1[0], v2[1] - v1[1], v2[2] - v1[2]};
```

Comment:
Can you think of a way to force a compile failure if `v1` isn't a 3-element container?

Reply:
I could add a remember statement like this:

Remember (std::array<double, 3> check_v1{v1};)
Require(v1.size()==3);

Reply:
I decided on just the Check(v1.size()==3). This will at least ensure it is a vector or array and in debug will through should it get passed one that is the wrong size.